core/store/src/db/refcount.rs

```diff
@@ -78,6 +78,13 @@ pub(crate) fn encode_value_with_rc(data: &[u8], rc: i64) -> Vec<u8> {
     cursor.into_inner()
 }
 
+pub(crate) fn encode_value_with_rc_inplace(
+    data: &mut Vec<u8>,
+    rc: i64,
+) -> Result<(), std::io::Error> {
```

Comment:
There's no way this can fail with `io::Error`, this should return `()`.

Meta note: "can operation X fail in a way Y" is an important low-level coding concern which should be kept in mind. It's too easy to propagate errors which do not make sense, so it's important to be vigilant and notice when error management does't make sense. This is the case where static types make the life harder, because they describe the situation which can not actually arise in practice.

Reply:
Can we get rid of `encode_value_with_rc` altogether? In general, it is idiomatic to let caller pass the buffers in, rather than allocate Vecs in callee.

Reply:
I think you mixed `rc` and `memory_usage` here. The `RawTrieNodeWithSize` struct contains `memory_usage`, which records the size of the node, not the ref count.

Ref count is stored in the lower database layer, see `merge_refcounted_records`.